### ISSUE TITLE
typeorm_metadata table not in migrations when generating view sql

### ISSUE THREAD

**Issue by timojokinen** [typeorm/typeorm#4923]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `cockroachdb`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`
[ ] `expo`

**TypeORM version:**

[x] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

**Steps to reproduce or a small repository showing the problem:**
When generating the SQL for a view with the cli command "migration:generate" the command creates a table called typeorm_metadata in the database without adding it to the migrations file. If you drop the schema and run the migrations again or run the migration on another machine it fails because the table doesn't exist.

A solution would be if the migration:run command creates the typeorm_metadata table if it doesn't exist like migration:generate does or if the migration:generate command adds the create SQL script to the migration.

A workaround is to add the sql create script for typeorm_metadata to your migrations manually.

Up:
``CREATE TABLE `typeorm_metadata` (`type` varchar(255) NOT NULL,`database` varchar(255) DEFAULT NULL,`schema` varchar(255) DEFAULT NULL,`table` varchar(255) DEFAULT NULL,`name` varchar(255) DEFAULT NULL,`value` text) ENGINE=InnoDB``

Down:
DROP TABLE `typeorm_metadata`

**Comment by pablosproject** [comment]
Thanks, I just banged my head against this.

**Comment by zaro** [comment]
Same here. What would make most sense I think is to move the creation of the table on run, instead of generate.

**Comment by timojokinen** [comment]
@zaro There's an open PR that addresses this

**Comment by jtushman** [comment]
Just ran into this as well.  Thanks for the work around.

**Comment by moltar** [comment]
Same query for PG:

```sql
            CREATE TABLE typeorm_metadata (
                "type" varchar(255) NOT NULL,
                "database" varchar(255) DEFAULT NULL,
                "schema" varchar(255) DEFAULT NULL,
                "table" varchar(255) DEFAULT NULL,
                "name" varchar(255) DEFAULT NULL,
                "value" text
            )
```

**Comment by niccolofanton** [comment]
I'm having the same issue, not sure if it's already solved by [this PR](https://github.com/typeorm/typeorm/pull/4726)  
I think [this ](https://github.com/typeorm/typeorm/issues/4123) could be the same issue 🤔

**Comment by madeindjs** [comment]
I had the same issue on Postgres with TypeORM 0.2.25.

**Comment by CesarLanderos** [comment]
Same in `0.2.25`

**Comment by wi-ski** [comment]
Any movement on this? Feels like a decent booboo

**Comment by andrei9669** [comment]
still persists

**Comment by gkTim** [comment]
We also ran into this issue

**Comment by michaelsogos** [comment]
Dear @AlexMesser 

Any news about this?
As described above, we have a local database server where developers make patches.
Recently we starts use VIEW and when we make a new migration script (via CLI) we will have an _INSERT INTO typeorm_metadata_ inside the generated patch.
The problem is that when we run patches on central server (for QA) it fails because the table never will be generated.

As suggested above the _CREATE TABLE_ should be included in generated patch; but let me ask you... do you really need that kind of metadata table? Which is its target?

**Comment by imnotjames** [comment]
> but let me ask you... do you really need that kind of metadata table? Which is its target?

Yes, we do.  This is because we cannot gather the original query for the view from the database itself in many cases.

**Comment by michaelsogos** [comment]
Dear @imnotjames

Maybe I can help you with this.
On which engine you cannot retrieve view definition?

If you agree, I suggest to move the responsability to retrieve view definition on driver specific implementation, in order to not have a metatable when it is not needed.

**Comment by imnotjames** [comment]
>  On which engine you cannot retrieve view definition?

Postgres, for example, does not support retrieving the original SQL.

**Comment by haggholm** [comment]
> Postgres, for example, does not support retrieving the original SQL. At least, it didn't when this feature was implemented.

https://www.postgresql.org/docs/current/functions-info.html

`pg_get_viewdef()` seems to have been available since version 8.0 (first released in 2005, 16½ years ago).

**Comment by imnotjames** [comment]
Pretty certain that it gives you back DDL that's been processed - not the original SQL.

**Comment by michaelsogos** [comment]
Dear @imnotjames

I'm the author of pg-diff-cli and its counter part pg-diff-api.
You are right, it will give you back a different DDL than one you submit; at the same time, based on my experience, you should avoid the approach to check differences between DDL you submit and real view definition re-interpreted by the engine; I'd like to suggest to keep final view definition directly on migration patch generated.
How?
You must execute statement defined in ViewEntity expression in order to create view on target db, then retrieve its final view definition like above comment, keep it on migration patch, then remove the view from database or rollback in some how.
In this way next comparison keep you safe and metatable isn't needed anymore.

Anyway, the problem here is to keep the CREATE metatable inside the migration patch.
Let eventually improve the approach for a different time 😄

**Comment by pleerock** [comment]
There is an opened PR that should fix this issue: #4956

It will be really helpful if you can review that PR, try it on your project, check if it works for you and approve PR. This way everyone can be sure PR does the right thing, it will be merged and you'll have issue fixed in your project.

P.S.: I'm addressing people who left "I have this issue too" comment. 

Thank you.

**Comment by lbguilherme** [comment]
This can be closed as the PR has been merged and released already.
